### PR TITLE
Patch For Modules

### PR DESCRIPTION
**Pull request by Zildj1an**
* I have exported the strictly necessary functions for developing a scheduler plugin from a kernel module (see #8)

* Also I added the function unregister_sched_plugin() for safely removing the uploaded module from the list, and included it in all the pre-made modules.

* The developer can maintain a reference counter as I did in [this module](https://github.com/Zildj1an/P-EDF_kernel_sched_althm/blob/master/mi_plugin.c) to safely upload the plugin.

Note:
There is also a detail I had to take into account regarding the usage of THIS_MODULE for the pre-made plugins. As can be seen in [export.h](https://github.com/LITMUS-RT/litmus-rt/blob/linux-4.9-litmus/include/linux/export.h) if the source file is compiled into the kernel rather than uploaded like a typical module, THIS_MODULE is set to NULL, which caused a kernel panic when I tried to boot from the new version. This is why the special pre-made plugins need to check if they are been loaded when booting (#ifdef MODULE) and therefore assume it has already been loaded once (when booting) when uploading again or been removed.
But the developer does not need to worry about this special case and can simply use THIS_MODULE like I did above.

### CODE REVIEW

**Comment by brandenburg** [comment]
Thanks for the patch!

I don't quite get the need for the manual counts and the `num_references` parameter that's being passed to register_sched_plugin() — what's up with this? AFAIK, Linux's register/unregister APIs don't typically have such parameters, so this is a non-obvious pattern to me. And what do you mean by  "uploaded several times"? Modules should be loaded only once. Could you please elaborate a bit?

**Comment by Zildj1an** [comment]
Hello,

The manual counter would not be necessary (the modules could simply use the ```module_put()``` and ```try_module_get()```), unfortunately those functions require a pointer to the module (given by THIS_MODULE) but if compiling the files with the kernel that value is set to NULL(line 33[ here](https://github.com/LITMUS-RT/litmus-rt/blob/linux-4.9-litmus/include/linux/export.h)) and when the OS boots it gets a kernel panic.

That is why I required a manual counter for the special pre-made plugins. For any other new plugin uploaded as a module that won't be a problem to the developer.

Regarding the parameter ```num_references``` I just thought it would be interesting if those functions where in charge of checking if the plugin has already been registered (with the number), before adding them to the list.

In this way, if several elements in the computer are using the module (they tried to upload it, that's what I meant by 'uploaded several times'), the module will not be unloaded until all of them attempt to unload it.

I don't know if this pattern is typical in the kernel, but it definitely works. Do you think it would be better to make the plugins itself check the counter before calling the functions register and unregister? I would say this approach is safer and reduces the work to the module developer, but if you think it is better the other way ... (I don't know if there may be a more appropriate pattern for this case)

Thank you!

**Comment by brandenburg** [comment]
Thanks for the explanation. The LITMUS^RT project generally strives for its code to look as much "Linux-like" as possible. In this light, I'd rather not introduce "interesting" new patterns like the `num_references` parameter.

In terms of guiding examples, I think file systems would be a great template: most Linux file systems can be built both as a module or as part of the core kernel, and they need to register/deregister a "plugin" (= a struct full of FS implementation callbacks) just like the policy plugins in LITMUS^RT. They pretty much must solve all the same problems, so let's emulate the same solutions.

**Comment by Zildj1an** [comment]
For what I can see in [fs/filesystems.c](https://github.com/LITMUS-RT/litmus-rt/blob/4acc377593580e7d04ad8b42b258e8c2b39030ee/fs/filesystems.c) between lines 69 and 120 what the kernel does is simply traverse the list to see if the fs already exists.
When registering, if already exists it does nothing.
When unregistering, if exists it is deleted.
 
That will make unnecessary the usage of a counter, we could do that for our register/unregister of the plugin. It is much simpler and Linux kernel-alike.

But that way the module could be unloaded when someone else is using it (because there is no counter of references), so I don't know if it's worth-it sticking to the typical kernel pattern for this case. Which one is better?

Also, other possible solution would be to add a field in the sched_plugin as counter of references. And check it whenever it is been inserted or removed. That way we could get rid of the parameter ```num_references```.

Which approach should I commit?

Regards!

**Comment by brandenburg** [comment]
I think you can simply prevent the currently active plugin from being removed. Since we don't allow plugin switching while there are active real-time tasks, any non-active plugin is going to be unused.

**Comment by Zildj1an** [comment]
Done! I have booted from this new version without problems. I have also tried loading and unloading my module plugin and it worked.

File: litmus/sched_cedf.c [hunk 1]
```diff
@@ -879,11 +880,13 @@ static int __init init_cedf(void)
 
 static void clean_cedf(void)
 {
-	cleanup_cedf();
-	if (cluster_file)
-		remove_proc_entry("cluster", cedf_dir);
-	if (cedf_dir)
-		remove_plugin_proc_dir(&cedf_plugin);
+	if(unregister_sched_plugin(&cedf_plugin)){
```
Comment (brandenburg) [inline litmus/sched_cedf.c]: please follow Linux coding conventions (spacing)

File: litmus/sched_cedf.c [hunk 2]
```diff
@@ -867,7 +867,8 @@ static int __init init_cedf(void)
 	int err, fs;
 
 	err = register_sched_plugin(&cedf_plugin);
-	if (!err) {
+
+	if (!err){
```
Comment (brandenburg) [inline litmus/sched_cedf.c]: please follow Linux coding conventions (spacing)

**Comment by brandenburg** [comment]
Thanks a lot for the revision. Before merging this, could you please make a pass through the diff to make the new code follow Linux coding style in all places? Thanks.

**Comment by Zildj1an** [comment]
My pleasure!
I took care of the spacing. 
Also I read [here](https://www.kernel.org/doc/html/v4.10/process/coding-style.html) that the EXPORT macro should -by convention- be placed immediately after the closing function brace line, so I have changed that. 
Thanks.

File: litmus/sched_pfair.c [hunk 3]
```diff
@@ -1224,6 +1224,7 @@ static int __init init_pfair(void)
 static void __exit clean_pfair(void)
 {
 	if(unregister_sched_plugin(&pfair_plugin)){
+
```
Comment (brandenburg) [inline litmus/sched_pfair.c]: Sorry, I think there has been a misunderstanding. I certainly didn't mean to suggest that you need a blank line after every 'if' statement. Rather, provide some space around operands, etc. Don't write it as:

```
if(foo){
```

It should instead be:

```
if (foo) {
```

Reference: https://github.com/LITMUS-RT/litmus-rt/blob/linux-4.9-litmus/Documentation/CodingStyle#L203

**Comment by Zildj1an** [comment]
Sorry for the misunderstanding, I believe it is correctly spaced now.
Thanks!

**Comment by brandenburg** [comment]
Looks great, thanks a lot for your attention to detail. 

Looking at the patch, I have one more question: I still don't understand how it prevents a module with a scheduler plugin from being unloaded while the plugin is in use. The diff shows that you removed the FIXME about obtaining a reference to the module that holds a scheduler when it is being registered, but I don't actually see any code that does this. What am I missing?

File: litmus/sched_plugin.c [hunk 4]
```diff
@@ -212,9 +213,22 @@ static DEFINE_RAW_SPINLOCK(sched_plugins_lock);
 	if (!plugin->func) \
 		plugin->func = litmus_dummy_ ## func;}
 
-/* FIXME: get reference to module  */
```
Comment (brandenburg) [inline litmus/sched_plugin.c:215]: This is the FIXME that I mean — the code should make sure that the module's reference count does not go to zero while the scheduler is in use, lest the active scheduler is being unloaded.

File: litmus/sched_plugin.c [hunk 5]
```diff
@@ -253,11 +267,30 @@ int register_sched_plugin(struct sched_plugin* plugin)
...
+
+	if (strcmp(litmus->plugin_name, plugin->plugin_name) != 0)
+	{
+		unregister = 1;
+		raw_spin_lock(&sched_plugins_lock);
+		list_del(&plugin->list);
+		raw_spin_unlock(&sched_plugins_lock);
+	}
+	else
+		printk(KERN_ALERT "The currently active plugin can not be removed\n");
```
Comment (brandenburg) [inline litmus/sched_plugin.c:287]: So what happens if this case is triggered during a module_exit() callback? How can it prevent the unloading at this point?

File: litmus/sched_pres.c [hunk 6]
```diff
@@ -608,5 +609,11 @@ static int __init init_pres(void)
 	return register_sched_plugin(&pres_plugin);
 }
 
-module_init(init_pres);
+static void __exit exit_pres(void)
+{	
+	if (unregister_sched_plugin(&pres_plugin))
```
Comment (brandenburg) [inline litmus/sched_pres.c:617]: What if this call fails because PRES is the active plugin and there are active real-time tasks?

**Comment by Zildj1an** [comment]
Hello,
I erased the part of counting the module's references because it was my understanding that since you don't allow plugin switching while there are active real-time tasks, any non-active plugin is going to be unused (and because the pattern I used if you remember was not linux-alike, with the extra parameter at ```register_sched_plugin``` and unregister functions).

Therefore you are right, hitherto there is no distinction between if it is not used and it has been loaded several times or only once, because it is just not added to the list again. I deleted that line because since it is not used nor added to the list there is no practical difference, AFAIK.

But if you think it would be nice to make it check the counter before erasing it from the list, I can use a linux-alike pattern for the reference counting. What if I add a variable at the scheduling plugin with the number of references, and check it and update it when attempting to load/unload it? Would that be a good approach?

Regards

File: litmus/sched_plugin.c (hunk 4 above)
Comment (Zildj1an) [reply litmus/sched_plugin.c:215]: Answered at the pull request conversation

File: litmus/sched_plugin.c (hunk 5 above)
Comment (Zildj1an) [reply litmus/sched_plugin.c:287]: Answered at the pull request conversation

File: litmus/sched_pres.c (hunk 6 above)
Comment (Zildj1an) [reply litmus/sched_pres.c:617]: Answered at the pull request conversation

**Comment by brandenburg** [comment]
My understanding is that there is a standard way to prevent Linux modules from being unloaded (e.g., think about the USB core being unloaded while there are still USB devices active). LITMUS^RT should be using that interface and not rely on a custom (and likely redundant) reference count. So please don't add a variable to each plugin manually, but figure out how to use the appropriate Linux API. Also, I don't think you can load a Linux module multiple times (what does it mean to load, e.g., ext2 multiple times?), so repeated registrations of scheduling policy plugins are not a pressing problem.

**Comment by Zildj1an** [comment]
I have seen that Linux has the struct kref that it is used for cases of multiple simultaneous use, as explained [here](https://www.mjmwired.net/kernel/Documentation/kref.txt), it is commonly used for USB drivers. 

And as you said, ```insmod``` (directly or from ``modprobe``) won't call init_module if it is already loaded, so this kref for the drivers is updated when the syscall open() for that module is called, and decreased when close() is invoked.

But, if it can not be loaded more than once, and those modules don't have open or close syscalls, is this really necessary?

**Comment by brandenburg** [comment]
What about this sequence of events:

1. insmod MYPOLICY.ko
2. setsched MYPOLICY
3. rtspin 10 100 100 &
4. rmmod MYPOLICY

What prevents the removal of the module in step 4? If it succeeds, the next time the scheduler is triggered, it will branch to functions in now-unallocated memory.

**Comment by Zildj1an** [comment]
A way of solving this could be:
When rtspin is called it increases a kref counter at the plugin.
When ```exit_module()``` is called it waits until this counter is zero.
 
But maybe a more stylish way to do this could be by adding a Linux mutex at the plugin, locked when calling rtspin, and make ```exit_module()``` to get it.

**Comment by brandenburg** [comment]
In no place does the example trace switch a policy plugin *while tasks are active*. That's exactly the point — yes, we prevent switching of the plugin while there are active RT tasks, but currently there is nothing to prevent the unloading of a module holding the plugin *while it is still active*. The need to address this problem is exactly why we never implemented support for policy plugins in (un-)loadable kernel modules. 

Can you actually wait in `exit_module()`? A good solution IMO would be to increase the module's reference count when the policy plugin is registered, and to decrease it again when it is being deregistered. How does this work with e.g. file systems and `mount`?

**Comment by Zildj1an** [comment]
UNMOUNT

Don't mind MNT_DETACH, is a flag to force the unmounting. In file systems Linux just check a counter of references (as you see, "not used anymore"), as we were guessing.

The problem with the solution you propose is that (I think) it does not solve:

> What about this sequence of events:
> 
>     1. insmod MYPOLICY.ko
> 
>     2. setsched MYPOLICY
> 
>     3. rtspin 10 100 100 &
> 
>     4. rmmod MYPOLICY
> 

The only disadvantage with the solution I propose (a mutex for the plugin) is that it wouldn't allow the simultaneous usage of the plugin, but this could be solved by using a counter kref instead, like the file systems.

**Comment by Zildj1an** [comment]
Ok now that I think about it we are saying almost the same thing but I believe the counter should not be increased/decreased at registering/unregister but when the plugin is used/stop been used by any other third party

**Comment by brandenburg** [comment]
Wrapping the entire plugin in a mutex is going to cause lockdep pain. I would not recommend that. 

I think the solution is simple: increment the module's use count when activating the policy (with `setsched`), and decrement it when deactivating it (again, with `setsched`). While active, it will be impossible to unload the module. 

I haven't looked into this in ages, but I believe the proper API to use is `try_module_get()` and `module_put()`.

**Comment by Zildj1an** [comment]
I'm sorry, I'm not sure if I'm following. Are you talking about modifying [setsched](https://github.com/LITMUS-RT/liblitmus/blob/a430c7b5cacae4219a185decb863f6fb9e5cbcb9/setsched) bash code to increase/decrease there the counter? So that ```rmmod``` will return -EBUSY and the only way to switch would be to call setsched for a different plugin the number of times it was the previous one set?

**Comment by brandenburg** [comment]
No. This is certainly not something that should be done in userspace; the kernel needs to ensure its integrity by itself. 

The code that you edited had some FIXMEs about getting a reference to the module holding the policy plugin. That's approximately the right place to manage the module's reference counter. Getting a reference = `try_module_get()`; dropping a reference = `module_put()`.

**Comment by Zildj1an** [comment]
I see.
What I can do is to modify the functions ```activate_plugin()``` and ```deactivate_plugin()``` of the pre-made plugins so they increase or decrease the modules reference counter. 

```exit_module()``` could then call ```deactivate_plugin()```.

But those are changes at the plugins, not at the register or unregister, that is only done once.

**Comment by Zildj1an** [comment]
Having the sequence:
```
  1  insmod MYPOLICY.ko
  2  setsched MYPOLICY
  3  rtspin 10 100 100 &
  4  rmmod MYPOLICY
```
will not be a problem if there is a counter of references when activating or deactivating the plugin, like I did in [my plugin](https://github.com/Zildj1an/P-EDF_kernel_sched_althm) with ``try_module_get()`` and ``module_put()``. 
If that counter has been implemented (as I did in the pre-made schedulers) there would be an error of module in use when trying to remove it until the current scheduler is changed.

(Compiled,boot and tried from this version without problem)

**Pull request closed without merging**